### PR TITLE
add train test split tools

### PR DESCRIPTION
**Pull request by qiagu**
The tool seems to be useful for certain circumstance.

### CODE REVIEW

**Comment by qiagu** [comment]
That `table_compute` is one of my favorite tools, but train_test_split does some particular tasks that perform `shuffle` split, `stratifiedShuffle` Split` and `groupShuffle` split as well. I don't think the `table_compute` supports all of them.

**Comment by qiagu** [comment]
I couldn't understand the failure report generated by the travis lint.

File: tools/sklearn/fitted_model_eval.xml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        <import>keras_macros.xml</import>
+    </macros>
+    <expand macro="python_requirements"/>
+    <expand macro="macro_stdio"/>
+    <version_command>echo "@VERSION@"</version_command>
+    <command>
+        <![CDATA[
+        export HDF5_USE_FILE_LOCKING='FALSE';
+        python '$__tool_directory__/fitted_model_eval.py'
+            --inputs '$inputs'
```
Comment (simonbray) [inline tools/sklearn/fitted_model_eval.xml]: Why do you feed some inputs in via a config file and others not?

File: tools/sklearn/fitted_model_eval.xml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            <param name="header2" value="true"/>
+            <param name="col2" value="1"/>
+            <output name="outfile_eval" file="fitted_model_eval01.tabular"/>
+        </test>
+    </tests>
+    <help>
+        <![CDATA[
+**What it does**
+
+Given a fitted estimator and a labeled dataset, this tool outpus the performances of the fitted estimator on the labeled dataset with the selected scorers.
```
Comment (simonbray) [inline tools/sklearn/fitted_model_eval.xml]:
```suggestion
Given a fitted estimator and a labeled dataset, this tool outputs the performances of the fitted estimator on the labeled dataset with the selected scorers.
```

File: tools/sklearn/fitted_model_eval.xml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+For estimator, this tool supports fitted sklearn estimators (pickled) and trained deep learning models (model skeleton + weights). For input datasets, it supports  as follows.
```
Comment (simonbray) [inline tools/sklearn/fitted_model_eval.xml]:
```suggestion
For an estimator, this tool supports fitted sklearn estimators (pickled) and trained deep learning models (model skeleton + weights). For input datasets, it supports the following:
```

File: tools/sklearn/stacking_ensembles.py [hunk 4]
```diff
@@ -82,7 +82,9 @@ def main(inputs_path, output_obj, base_paths=None, meta_path=None,
 
     weights = options.pop('weights', None)
     if weights:
-        options['weights'] = ast.literal_eval(weights)
+        weights = ast.literal_eval(weights)
+        if weights:
+            options['weights'] = weights
```
Comment (simonbray) [inline tools/sklearn/stacking_ensembles.py]: do you also want to bump the version number in the XML file so the tool is updated?

File: tools/sklearn/train_test_split.xml [hunk 5]
```diff
@@ -0,0 +1,115 @@
+<tool id="sklearn_train_test_split" name="Split Dataset" version="@VERSION@">
+    <description>into train and test subsets</description>
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]:
```suggestion
    <description>into training and test subsets</description>
```

File: tools/sklearn/train_test_split.xml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        <import>main_macros.xml</import>
+    </macros>
+    <expand macro="python_requirements"/>
+    <expand macro="macro_stdio"/>
+    <version_command>echo "@VERSION@"</version_command>
+    <command detect_errors="exit_code"><![CDATA[
+        python '$__tool_directory__/train_test_split.py'
+            --inputs '$inputs'
+            --infile_array '$infile_array'
+            #if $options.shuffle_selection.shuffle not in ['None', 'simple']
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]:
```suggestion
            #if $options.shuffle_selection.shuffle not in ['None', 'simple']:
```

File: tools/sklearn/train_test_split.xml [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            --infile_labels '$options.shuffle_selection.labels'
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]:
```suggestion
                --infile_labels '$options.shuffle_selection.labels'
```

File: tools/sklearn/train_test_split.xml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            --outfile_test '$out_test'
+    ]]>
+    </command>
+    <configfiles>
+        <inputs name="inputs"/>
+    </configfiles>
+    <inputs>
+        <param name="infile_array" type="data" format="tabular" label="Select the dataset containing array to split"/>
+        <param name="infile_info" type="select" multiple="true" display="checkboxes" optional="true" label="The dataset">
+            <option value="header" selected="true">Has headers</option>
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]: Why not simply a boolean here?

File: tools/sklearn/train_test_split.xml [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        </param>
+        <section name="options" title="Options" expanded="true">
+            <param argument="test_size" type="float" optional="False" value="0.25" label="Test size:"
+                help="If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the test split. If int, represents the absolute number of test samples."/>
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]: I think this is not very intuitive. Could you create an option to allow the user to choose to enter either an absolute value or a proportion?

File: tools/sklearn/train_test_split.xml [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            --infile_array '$infile_array'
+            #if $options.shuffle_selection.shuffle not in ['None', 'simple']
+            --infile_labels '$options.shuffle_selection.labels'
+            #end if
+            --outfile_train '$out_train'
+            --outfile_test '$out_test'
+    ]]>
+    </command>
+    <configfiles>
+        <inputs name="inputs"/>
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]: same here as for the other file

File: tools/sklearn/train_test_split.xml [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+            <conditional name="shuffle_selection">
+                <param name="labels" value="regression_y.tabular" ftype="tabular"/>
+                <param name="infile_info" value="header"/>
+                <param name="col" value="1"/>
+            </conditional>
+            <output name="out_train" file="train_test_split_train03.tabular" ftype="tabular"/>
+            <output name="out_test" file="train_test_split_test03.tabular" ftype="tabular"/>
+        </test>
+    </tests>
+    <help><![CDATA[
```
Comment (simonbray) [inline tools/sklearn/train_test_split.xml]: You could expand this a little.

File: tools/sklearn/train_test_split.xml (hunk 8 above)
Comment (simonbray) [reply tools/sklearn/train_test_split.xml]: and for the other params following

**Comment by simonbray** [comment]
ping @qiagu - are you still working on this? Would be nice to get it merged. :)

File: tools/sklearn/fitted_model_eval.xml (hunk 1 above)
Comment (qiagu) [reply tools/sklearn/fitted_model_eval.xml]: Due to a Galaxy bug, the config file was unable to include input file paths. A good news is that the bug got fixed a few days ago.

File: tools/sklearn/stacking_ensembles.py (hunk 4 above)
Comment (qiagu) [reply tools/sklearn/stacking_ensembles.py]: This is a simple bug fix. Is backporting possible?

File: tools/sklearn/train_test_split.xml (hunk 6 above)
Comment (qiagu) [reply tools/sklearn/train_test_split.xml]: I have noticed that having or not having  the `:` here were the same.

File: tools/sklearn/train_test_split.xml (hunk 9 above)
Comment (qiagu) [reply tools/sklearn/train_test_split.xml]: I agree with you. yet, do you notice that scikit-learn sets the parameter like that? Maybe most people just stick to the proportion one.

File: tools/sklearn/train_test_split.xml (hunk 10 above)
Comment (qiagu) [reply tools/sklearn/train_test_split.xml]: ? Sorry.

File: tools/sklearn/train_test_split.xml (hunk 5 above)
Comment (qiagu) [reply tools/sklearn/train_test_split.xml]: Thanks for all the English corrections.

File: tools/sklearn/stacking_ensembles.py (hunk 4 above)
Comment (simonbray) [reply tools/sklearn/stacking_ensembles.py]: It is, but requires a manual update of the tool on usegalaxy.eu by @bgruening. I usually find it simpler to bump the version number also for small changes.

File: tools/sklearn/train_test_split.xml (hunk 10 above)
Comment (simonbray) [reply tools/sklearn/train_test_split.xml]: I meant regarding the inputs in a config file.

**Comment by qiagu** [comment]
I have added what I planned long time ago for the train test split tool and many concerns from review were addressed. If there is anything I missed, please don't hesitate to let me know.  A simple `model_fit` tool is on my target, but that could go to a different PR.
